### PR TITLE
Fix incorrect usage of arb.smap in tests

### PR DESCRIPTION
**Pull request by Avaq**
As I understand, the .smap method is not a map function. Its intended use
is to provide a way for JSVerify to verify that, given a set, a reduction
of that set through a given function will produce a succesful assertion if
the greater set fails the same assertion.

It takes two functions: f, and g, where g should be the right inverse of
f. The current code only passes f, and leaves g to undefined. This means
that if JSVerify would ever find a failing property using one of these
incorrectly smapped arbitraries, and deem it necessary to attempt
shrinking it, it would throw an error like: 'g is not a function'.

See also https://github.com/jsverify/jsverify/issues/104

This commit makes two changes:

1. A 'smapArb' dispatcher was added which forces both arguments to be
   provided. All occurrences of the arb.smap call where a right inverse
   function could be defined have been replaced to use this new
   dispatcher.
2. A 'mapArb' dispatcher was added which can be used safely in the way
   that arb.smap was previously used. It relies on jsc.bless to attach a
   default shrink function. The default shrink function just produces an
   empty set.

<sub>/cc @dicearr (him and I have been bitten by this same mistake many times in the past)</sub>

### CODE REVIEW

**Comment by davidchambers** [comment]
> This means that if JSVerify would ever find a failing property using one of these incorrectly smapped arbitraries, and deem it necessary to attempt shrinking it, it would throw an error like: 'g is not a function'.

I have encountered this many times! Thank you for fixing the underlying issue. :clap:

File: test/index.js [hunk 1]
```diff
@@ -23,24 +24,23 @@ const {withUnstableArraySort} = require ('./quicksort');
 const {Nil, Cons} = List;
 const {Nothing, Just} = Maybe;
 
+//    mapArb :: (a -> b) -> Arbitrary a -> Arbitrary b
```
Comment (davidchambers) [inline test/index.js:27]:
```suggestion
//  mapArb :: (a -> b) -> Arbitrary a -> Arbitrary b
```

File: test/index.js [hunk 2]
```diff
@@ -23,24 +24,23 @@ const {withUnstableArraySort} = require ('./quicksort');
...
+const mapArb = f => arb => jsc.bless ({generator: arb.generator.map (f)});
+
+//    smapArb :: (a -> b) -> (b -> a) -> Arbitrary a -> Arbitrary b
```
Comment (davidchambers) [inline test/index.js:30]:
```suggestion
//  smapArb :: (a -> b) -> (b -> a) -> Arbitrary a -> Arbitrary b
```

File: test/index.js [hunk 3]
```diff
@@ -77,6 +77,13 @@ function args() {
   return arguments;
 }
 
+//    arrayToList :: Array a -> List a
```
Comment (davidchambers) [inline test/index.js:80]:
```suggestion
//  arrayToList :: Array a -> List a
```

File: test/index.js [hunk 4]
```diff
@@ -1652,15 +1659,19 @@ suite ('laws', () => {
   });
 
   suite ('RegExp', () => {
-    const arb = jsc.string.smap (s => {
-      const validFlags = ['', 'g', 'i', 'm', 'gi', 'gm', 'im', 'gim'];
-      const flags = validFlags[Math.floor (Math.random () * validFlags.length)];
```
Comment (Avaq) [inline test/index.js:1657]: While I was at it, I also removed this problematic code. One of the benefits of prop tests is that you can reproduce a failing assertion by rerunning the tests with the same seed. Of course this depends on your arbitraries being defined in terms of pure functions.

File: test/index.js [hunk 5]
```diff
@@ -1652,15 +1659,19 @@ suite ('laws', () => {
...
+    const flagsArb = jsc.oneof (['', 'g', 'i', 'm', 'gi', 'gm', 'im', 'gim'].map (jsc.constant));
+    const patternArb = jsc.suchthat (jsc.string, s => {
       try {
-        return new RegExp (s, flags);
-      } catch (err) {
-        return /(?:)/;
```
Comment (davidchambers) [inline test/index.js:1661]: Is `/(?:)/` no longer necessary because `jsc.suchthat` rejects strings that are not regular expressions?

Comment (Avaq) [reply test/index.js:1661]: Yes.

File: test/index.js [hunk 6]
```diff
@@ -23,24 +24,23 @@ const {withUnstableArraySort} = require ('./quicksort');
...
-    let list = Nil;
-    for (let idx = 0; idx < xs.length; idx += 1) list = Cons (xs[idx], list);
-    return list;
-  })
-);
+const ListArb = arb => smapArb (arrayToList) (listToArray) (jsc.array (arb));
 
 //  MaybeArb :: Arbitrary a -> Arbitrary (Maybe a)
-const MaybeArb = arb => jsc.oneof (jsc.constant (Nothing), arb.smap (Just));
+const MaybeArb = arb => jsc.oneof (jsc.constant (Nothing), mapArb (Just) (arb));
```
Comment (Avaq) [inline test/index.js:43]: It might be worth adding a custom shrink function here to get back to an `Arb a` given an `Arb (Maybe a)`. :thinking:

Comment (Avaq) [reply test/index.js:43]: Hm. No. :P

**Comment by Avaq** [comment]
@davidchambers Why did the tests fail?

File: test/index.js [hunk 7]
```diff
@@ -77,6 +77,13 @@ function args() {
   return arguments;
 }
 
+//  arrayToList :: Array a -> List a
+function arrayToList(xs) {
+  let list = Nil;
+  for (let idx = 0; idx < xs.length; idx += 1) list = Cons (xs[idx], list);
+  return list;
+};
```
Comment (davidchambers) [inline test/index.js:85]: Let's use an arrow function.

We use function declarations below to enable us to verify the correctness of implementations by checking `arguments.length` (`this.map (f)` is not an acceptable implementation of `Array#fantasy-land/map` due to the extra arguments). Such assertions are not required here, so we can use an arrow function. :)

**Review by davidchambers — APPROVED** [review]
:clap:

**Pull request merged**